### ISSUE TITLE
Problems using temporary credentials obtained from STS

### ISSUE THREAD

**Issue by craighurley** [OlafConijn/AwsOrganizationFormation#70]
Hello,

I like what I've read about `org-formation` but I'm having problems running it with my STS credentials.  Can you please help trouble shoot this issue?

I limit assuming roles to users that have MFA enabled.  I use [aws-mfa](https://github.com/broamski/aws-mfa) to manage obtaining temporary credentials from STS.

```
$ org-formation --version
0.9.4
```

My `~/.aws/config` is:

```
[profile ch-master]
region = eu-west-1

[profile ch-master-administrator]
region = eu-west-1
source_profile = ch-master
role_arn = arn:aws:iam::111111111111:role/Administrator
```

Initially, my `~/.aws/credentials` is:

```
[default]
aws_access_key_id = none
aws_secret_access_key = none

[ch-master-long-term]
aws_mfa_device = arn:aws:iam::111111111111:mfa/craig.hurley.cli
aws_access_key_id = AAAAAAAAAAAAAAAAAAAA
aws_secret_access_key = BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
```

Next, I run `aws-mfa --profile ch-master`, enter my MFA code at the prompt, STS returns the tokens and `aws-mfa` adds the following to my `~/.aws/credentials` file.

```
[ch-master]
assumed_role = False
aws_access_key_id = XXXXXXXXXXXXXXXXXXXX
aws_secret_access_key = yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy
aws_session_token = mmmmmmmmm
aws_security_token = nnnnnnnnn
expiration = 2020-05-09 23:28:10
```

At this point I can successfully run commands on the aws cli using the role defined under `[profile ch-master-administrator]` in my `~/.aws/config` file, e.g:

```
$ aws s3 ls --profile ch-master-administrator
2020-05-08 10:19:24 cf-templates-wxyw1t5ayxrv-eu-west-1
...
```

However, `org-formation` fails to find the profile (and use the STS provided tokens):

```
$ org-formation init org.yaml --verbose --print-stack --region eu-west-1 --profile ch-master-administrator
DEBG: initializing, arguments:
{
  "stateBucketName": "organization-formation-${AWS::AccountId}",
  "stateObject": "state.json",
  "state": "undefined",
  "profile": "ch-master-administrator",
  "color": true,
  "verbose": true,
  "printStack": true
}
👋 Enter MFA code for undefined:
112233
ERROR: unexpected error occurred...
Profile ch-master-administrator not found
SharedIniFileCredentialsProviderFailure: Profile ch-master-administrator not found
    at SharedIniFileCredentials.load (/usr/local/lib/node_modules/aws-organization-formation/node_modules/aws-sdk/lib/credentials/shared_ini_file_credentials.js:101:11)
    at SharedIniFileCredentials.coalesceRefresh (/usr/local/lib/node_modules/aws-organization-formation/node_modules/aws-sdk/lib/credentials.js:205:12)
    at SharedIniFileCredentials.refresh (/usr/local/lib/node_modules/aws-organization-formation/node_modules/aws-sdk/lib/credentials/shared_ini_file_credentials.js:168:10)
    at SharedIniFileCredentials.get (/usr/local/lib/node_modules/aws-organization-formation/node_modules/aws-sdk/lib/credentials.js:122:12)
    at resolveNext (/usr/local/lib/node_modules/aws-organization-formation/node_modules/aws-sdk/lib/credentials/credential_provider_chain.js:125:17)
    at /usr/local/lib/node_modules/aws-organization-formation/node_modules/aws-sdk/lib/credentials/credential_provider_chain.js:126:13
    at /usr/local/lib/node_modules/aws-organization-formation/node_modules/aws-sdk/lib/credentials.js:124:23
    at /usr/local/lib/node_modules/aws-organization-formation/dist/src/util/aws-util.js:153:29
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by OlafConijn** [comment]
Hi Craig, thanks for your issue!

I will look into this today or tomorrow a bit more thoroughly (will try to reproduce using the setup you describe. thanks for that).

If i understand you correctly you would expect not to get prompted for a MFA code after having ran `aws-mfa`. You would expect the temporary credentials written to `~/.aws/credentials` to be used directly.

For this case, indeed, the loading of credentials using a profile seems broken. From what i can tell:
1. The custom implementation for MFA part of org-formation seems to follow the try to do the mfa challange but fails
2. The SharedIniFileCredentials provider (from AWS) doesn't seem to find the profile.
 
One, obviously very lame workaround I can offer at this moment with v0.9.4 is to copy the temporary credentials tot the environment variables and not use a profile. This way you avoid using either method from above. how to do this is described here: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html 

Good news for now is that there is a really mature beta for v.0.9.5 that will be released soon and I am looking forward to release a fix for this issue as well.

Will keep you posted.
Olaf

**Comment by craighurley** [comment]
> you would expect not to get prompted for a MFA code after having ran aws-mfa. You would expect the temporary credentials written to ~/.aws/credentials to be used directly.

That’s right.

**Comment by OlafConijn** [comment]
Got it! 

fixed two issues:
1. Support for assumed credentials in the credentials file, as this requires the sessionToken to also be used.
2. Only asking for the MFA challenge using the tool if one, indeed, is configured. 

As a result it will now assume the role `arn:aws:iam::111111111111:role/Administrator` using the credentials generated by the `aws-mfa` tool for profile `ch-master`.

Thanks a lot for filing this issue and the clear description.
The commit this is fixed is in: https://github.com/OlafConijn/AwsOrganizationFormation/commit/306c7558af34cba09de1745543b1d174ab1ca940

Unfortunately the node sdk doesnt contain a credentials provider that supports mfa / assume role.

I will soon create a beta version and update this thread.
Olaf

**Comment by OlafConijn** [comment]
should be fixed in beta 17 of 0.9.5

`npm i aws-organization-formation@0.9.5-beta.17`

looking forward to hear whether this works!

**Comment by craighurley** [comment]
Version `0.9.5-beta.17` works for this scenario.  Thank you!

Well done on a great tool.

**Comment by OlafConijn** [comment]
released as part of v0.9.5
